common/flatpak-run.c

```diff
@@ -2963,8 +2963,46 @@ flatpak_run_app (const char     *app_ref,
 
   if (app_deploy != NULL)
     {
+      g_autofree const char **previous_ids = NULL;
+
       app_files = flatpak_deploy_get_files (app_deploy);
 
+      g_autoptr (GFile) current_data_dir = flatpak_get_data_dir (app_ref_parts[1]);
+      previous_ids = flatpak_deploy_data_get_previous_ids (app_deploy_data);
+      if (previous_ids)
+        {
+          guint len = g_strv_length ((char **) previous_ids);
+          while (len > 0)
+            {
+              g_autoptr (GFile) previous_data_dir = NULL;
+              g_autoptr (GError) local_error = NULL;
+
+              if (g_file_query_exists (current_data_dir, cancellable))
+                break;
+
+              previous_data_dir = flatpak_get_data_dir (previous_ids[--len]);
+              if (!g_file_query_exists (previous_data_dir, cancellable))
+                continue;
+
+              if (!flatpak_file_rename (previous_data_dir,
+                                        current_data_dir,
+                                        cancellable, &local_error))
+                 {
+                   g_propagate_error (error, g_steal_pointer (&local_error));
+                   return FALSE;
+                 }
+
+               if (!symlink (g_file_get_path (current_data_dir),
```

Comment:
Nitpick: Personally, I think it’s clearer to do `if (symlink (…) == 0)` to check for success, rather than `if (!symlink (…))`. The latter reads as “if not symlink”, which suggests failure to me.